### ISSUE TITLE
How to assign role via authentication to container component with wampcra

### ISSUE THREAD

**Issue by antonymott** [crossbario/crossbar#1631]
Not clear how to assign role via authentication to components in containers with wampcra.
Docs aren't clear on this, and there is no example in docs or in crossbar-examples.authentication.

Issue#507 confirms that roles only valid for "in-router" components. For components in containers, assign role via authentication" https://github.com/crossbario/crossbar/issues/507
My workaround is that I put my container component app inside a router component, but I'd like to have container functionality.

Error message:
2019-09-17T14:48:33-0400 [Container     569] TLS client using platform trust
2019-09-17T14:48:33-0400 [Container     569] session closed with reason wamp.error.no_auth_method [cannot authenticate using any of the offered authmethods ['anonymous']]
2019-09-17T14:48:33-0400 [Container     569] Fatal error in component: While firing 'leave' event - Strings are not supported by Failure

To replicate, here is the relevant section of config.json

                    "paths": {
                        "/": {
                            "type": "static",
                            "directory": "../web"
                        },
                        "ws": {
                            "type": "websocket",
							"cookie": {
								"name": "cbtid",
								"length": 24,
								"max_age": 600,
								"store": {
									"type": "file",
									"filename": "cookies.dat"
								}
							},
                            "auth": {
                                "wampcra": {
                                    "type": "dynamic",
                                    "authenticator": "com.example.authenticate"
                                }
                            }
                        }
                    }
                }
            ],
            "components": [
                {
                    "type": "class",
                    "classname": "authenticator.AuthenticatorSession",
                    "realm": "realm1",
                    "role": "authenticator"
                },
                {
                    "type": "class",
                    "classname": "backend.BackendSession",
                    "realm": "realm1",
                    "role": "backend"
                }
            ]
        },
        {
            "type": "container",
            "options": {
                "pythonpath": [
                    ".."
                ]
            },
            "components": [
                {
                    "type": "class",
                    "classname": "db.AppSession",
                    "realm": "realm1",
                    "transport": {
                        "type": "websocket",
                        "endpoint": {
                            "type": "tcp",
                            "host": "127.0.0.1",
                            "port": 443,
                        	"tls": {
                        		"hostname": "localhost"
                        	}
                        },
                        "url": "wss://localhost:443/ws"
                    }
                }
            ]
        }
    ]
}

**Comment by antonymott** [comment]
Thanks Tobias for assigning meejah. If it turns out you can replicate and it is not just user-error :), I will happily offer to update the docs and code examples with the solution/code-clarification to help others in the future. I didn't hear from anyone at CB yet, but hopefully you got my signed ICAA so I can help out?

**Comment by meejah** [comment]
I will be able to look at this in a few hours; is your dynamic authenticator returning a `dict` with `"role": "custom_role"` or so in it?

**Comment by antonymott** [comment]
Thanks meejah. As start up progresses to worker001, from cb log: Site (TLS) starting on 443 static service (router.components), Ok, Transport transport001 has started Web Service webservice001. Web-inspector log confirms challenge response as expected: '{"authid": "peter", "authrole": "frontend", "authmethod": "wampcra", "authprovider": "dynamic"... 

All good so far, terminal log indicates: Ok, node has started Container worker002, TLS client using platform trust, session closed with reason wamp.error.no_auth_method [cannot authenticate using any of the offered authmethods ['anonymous']]. But web-inspector has no info about dict return object, just "connection lost".

One idea: my real (i.e. not home made) certificate doubles as a server cert and client cert, but the difference between my cert and CB example is that my CA-issued has only my FQDN, not "localhost" in the Alt DNS section. The mismatch caused me trouble to get the CB example container.component code to work before I tried adding wampcra. For example, even dropping the container.component tls options of cert, key and int-ca, it wouldn't work until I changed  container.component endpoint tls to "hostname: {my FQDN}". This change broke the autobahn.connection in index.html until I changed wsuri from localhost to the fqdn, but that makes index.html not so portable.

Once we get wampcra working, I plan to update the CB example with all that I have learned especially for real world cases where CB users have CA-issued client certs that like mine may not have an Alt DNS "localhost" option.

**Comment by meejah** [comment]
For testing / localhost it's possible to use a self-signed certificate; you have to ask the client to trust your issuing authority (i.e. the cert itself if it's actually self-signed).

In any case: updating the examples / docs would be great! (It's sometimes hard to know what's confusing when you already know the system).

If you're building a system from scratch, `cryptosign` auth might be better to look at (pubic/private keys instead of passwords).

**Comment by antonymott** [comment]
OK, thanks for tip on cryptosign. Regarding wampcra auth issue with container.component, I'm happy to first try your idea: I'll make a full set of: self-signed ca-root, intermediate root, server cert so I can augment the example with the functionality of static tls web which is a likely real world implementation, and also client cert for the container and get back to everyone with results...

**Comment by meejah** [comment]
@antonymott see here for that: https://github.com/crossbario/autobahn-python/blob/master/examples/twisted/wamp/pubsub/tls/create-self-signed-cert.sh

**Comment by meejah** [comment]
@antonymott does this example help? Here, the procedure called by dynamic auth is returning a dict -- because it has `"role"` in it, that role should be assigned to clients authenticating. This is "ticket" auth, but it should work the same for all types of dynamic auth (cryptosign, wampcra, etc).

https://github.com/crossbario/crossbar-examples/blob/master/authentication/advanced/authenticator.py#L74

I'm very interested to work with you to make this easier to discover from the documentation :) so maybe the best is -- once you figure out how to make your own code run, propose a doc-PR and I can review? (It might be useful to write down what you read and how you found it so we can make this stuff more discoverable).

Another avenue might be to add `wampcra` auth to the example above? (I would also like to update the clients in that example to use the `Component` API which makes authentication way easier, because you don't have to write `onChallenge` yourself).

**Comment by antonymott** [comment]
@meejah this was for me a tough problem to find, but I think I found the cause, if not the solution.

tl;dr
container.component client cert requires two certificate client cert attributes, one of them a deprecated Netscape extension attribute nsCertType=client. This double requirement of client certs is likely to break CB in production grade environments as deprecated extensions are not likely to be set by a CA unless specifically requested by the customer. My CA-signed cert does not have any Netscape extensions at all.

The longer version:
I found someone else (CB user) with a similar problem trying to move from self-signed example to CA example here https://github.com/crossbario/autobahn-python/issues/936. Tobias could not replicate, so closed. But not solved, so I pressed on.

I recreated sets of self-signed certs using the handy .sh script. For the testing environment, I temporarily short-circuited authenticator.py sha1 so any valid client.crt would allow crossbar to run as long as the tls steps down in the bowels of twisted and pyopenssl worked.

After I got things to work exactly like the example, I changed parameter by parameter of the self-signed client.crt to find which broke container.component.

I found:

client.crt must have extendedKeyUsage = clientAuth
otherwise: 
SSL error: sslv3 alert unsupported certificate (in ssl3_read_bytes)

This seems practical as most CAs issue certs with this attribute set conveniently to both clientAuth and serverAuth. My CA issued-cert did have this attribute set to both client and server, so all good there.

I also found:

client.cert requires deprecated Netscape extension nsCertType = client
otherwise:
SSL error: sslv3 alert unsupported certificate (in ssl3_read_bytes)

Probably should be a way to squash this error because it relies on the deprecated Netscape extension. Difficult to diagnose because it produces an identical error to the above. Reference to deprecated status: "Deprecated Netscape Extensions" https://www.openssl.org/docs/manmaster/man5/x509v3_config.html OpenSSL.org

Hard for CB users to diagnose: using the example self-signed cert which we know has nsCertType set, on Ubuntu 18.04.3 LTS, using "view" and reading every entry, no where does the nsCertType attribute appear. To even see nsCertType one has to run the openssl x509 command from terminal.

Is there a workaround?
I couldn't see an obvious workaround as I couldn't find the function/class that produces the error routine anywhere in crossbar, twisted or pyopenssl code. I did grep -r "sslv3 alert unsupported certificate" and found one match in a binary file ./site-packages/cryptography/hazmat/bindings/_openssl.abi3.so.

Could you please confirm if you think I'm on the right track and if there is a way to squash the nsCertType error after its propagated?

**Comment by antonymott** [comment]
@meejah: could the problem be that I'm trying to use the same CA-signed server cert as both server cert and client cert in three sections of config.json? I didn't think there should be a problem as my server cert has both uses defined: X509v3 Extended Key Usage: 
                TLS Web Client Authentication, TLS Web Server Authentication
In my config.json, I reference the one cert in:
transports (type:web),endpoint.tls
paths.ws.tls
and again in container.component: transport.endpoint.tls.

CB completes worker001 successfully, begins configuring worker002, loads TLS client cert, key and the ca_certificates, then hangs with message  [Container   17915] SSL error: certificate verify failed (in tls_process_server_certificate)

**Comment by meejah** [comment]
Thanks for all the testing work!

Since this is now sounding like some kind of certificate / TLS issue, can you confirm that you can make the dynamic-auth role assignment work (e.g. by using non-TLS connections) and then we can change the bug title to something more relevant ("client TLS ..."). Does that make sense, or is the wampcra/dynamic-auth still relevant?

**Comment by antonymott** [comment]
will do

**Comment by antonymott** [comment]
@meejah finally got it all working! Yay.

The current docs on github and crossbar.io show great working examples of tls (mostly self-signed), containers, and router apps, individually. I think others like me may be coming to crossbar who want a production grade implementation (ie certificate authority CAs) and who are trying to combine tls with router components, web components and in particular container.component transport tls from piecing together the doc examples.

If you think it might help others, I'm happy to turn this into a new doc example, or edit an existing example, to show how to make an implementation that uses commercial CA certificates, two router.components, wampcra on the websocket upgrade, web tls, websocket tls and a container.component thrown in for good measure...also using tls.

I'll separately go over errors and traps for the unwary - and how to avoid them - but perhaps the best place for that is in the docs. As an example, one of the biggest problems I found was tracing the cause of a set of tls errors: it was me not realizing we have to use chain_certificates array holding one or more intermediate certificate authorities in router transports (web, websocket) BUT ca_certificates array with intermediate certificates authorities and root authority in container.components. From google searches, I could see others in various forums over the last couple of years had the same or a similar problem, and importantly some of the threads were answered but not all. So perhaps this will help others.

Before I work on the docs and generate a pull request, can you (or Tobias or other crossbar expert) please look over my working .config extract below to let me know if it makes sense or if you'd like me to fix/improve it to be a better example:

relevant extract from config.json works. Yay.
{
    "version": 2,
    "controller": {},
    "workers": [
        {
            "type": "router",
            "options": {
                "pythonpath": [
                    ".."
                ]
            },
            "realms": [
                {
                    "name": "realm1",
                    "roles": [
                        {
                           ...
                        }
                    ]
                }
            ],
            "transports": [
                {
                    "type": "web",
                    "endpoint": {
                        "type": "tcp",
                        "port": 80
                    },
                    "paths": {
                        "/": {
                            "type": "redirect",
                            "url": "https://example.com"
                        }
                    }
                },
                {
                    "type": "websocket",
                    "endpoint": {
                        "type": "tcp",
                        "port": 9000,
						"tls": {
							"certificate": "/etc/ssl/certs/example_com.pem",
                        	"key": "/etc/ssl/private/example_com.key",
                        	"chain_certificates": [
                        		"/etc/ssl/certs/intermediate.pem"
                        		],
                        	"dhparam": "dhparam.pem"
						}
                    }
                },
                {
                    "type": "universal",
                    "endpoint": {
                        "type": "tcp",
                        "port": 443,
                        "tls": {
                        	"certificate": "/etc/ssl/certs/example_com.pem",
                        	"key": "/etc/ssl/private/example_com.key",
                        	"chain_certificates": [
                        		"/etc/ssl/certs/intermediate.pem"
                        		],
                        	"dhparam": "dhparam.pem"
                        }
                    },
                    "web": {
                    	"paths": {
		                    "/": {
		                        "type": "static",
		                        "directory": "../web"
		                    },
		                    "shared": {
		                        "type": "static",
		                        "directory": "/srv/_shared-web-resources"
		                    },
		                    "ws": {
		                        "type": "websocket",
		                        "auth": {
		                            "wampcra": {
		                                "type": "dynamic",
		                                "authenticator": "com.example.authenticate"
		                            }
		                        }
		                    }
		                }
                    }
                }
            ],
            "components": [
                {
                    "type": "class",
                    "classname": "authenticator.AuthenticatorSession",
                    "realm": "realm1",
                    "role": "authenticator"
                },
                {
                    "type": "class",
                    "classname": "backend.BackendSession",
                    "realm": "realm1",
                    "role": "backend"
                }
            ]
        },
        {
            "type": "container",
            "options": {
                "pythonpath": [
                    ".."
                ]
            },
            "components": [
                {
                    "type": "class",
                    "classname": "db.AppSession",
                    "realm": "realm1",
                    "transport": {
                        "type": "websocket",
                        "endpoint": {
                            "type": "tcp",
                            "host": "127.0.0.1",
                            "port": 9000,
                            "tls": {
                            	"hostname": "example.com",
                            	"certificate": "/etc/ssl/certs/example_com.pem",
                        	"key": "/etc/ssl/private/example_com.key",
                        	"ca_certificates": [
                        		"/etc/ssl/certs/intermediate.pem",
                        		"/etc/ssl/certs/rootCA.pem"
                        		]
                            }
                        },
                        "url": "wss://example.com:9000/ws"
                    }
                }
            ]
        }
    ]
}

**Comment by meejah** [comment]
Hi @antonymott I'm glad you got it working!

An improved or completely new example would be great. Also, if you have suggestions for messages which would have made this easier I think filing crossbar (or even autobahn) tickets for those would be great! I do agree that things can be hard to debug (but also turning on "debug" level logs can produce an overwhelming amount of information).

Thanks for digging into this!

**Comment by oberstet** [comment]
@antonymott good to hear you've got it working! and thanks for the nice words;)

> I think we have to stop our crossbar router before reloading an edited config, unless I've missed an important doc on dynamic config.

yes, crossbar open-source will read the node config exactly once when it starts, and never revisit the config while it is running.

crossbarfx (an extended version) does allow dynamic reconfiguration at run-time, without stopping a node, via a WAMP based management API .. let me know if you are interested in that ..

> Please let me know if a new doc example titled like that would be interesting to CB, or if you'd prefer I just edit one of the existing examples.

oh yes, that would be highly welcome! in general, I think our examples/docs would need even more love;) It has become a bit unwieldly, and as we have reached/explored the sensible scope of crossbar feature wise now, we're entering a phase of robustification, filling gaps in docs, correcting errors in docs and in general make all this stuff more approachable for newcomers!

so yes, your feedback, input and of course even PRs _are_ welcome! =)
